macros/timeHistoryAnalyzer.py

```diff
@@ -0,0 +1,457 @@
+
+            start = end + 1
+        else:
+            start += 1
+        pass
+
+    return ranges
+
+
+def findRangesMaskReason(data, vfat, channel, maxSkip = 5, minBadScans = 4):
+    """Finds ranges of scans based on the maskReason attribute.
+
+    Searches the data for the given vfat and channel for ranges of scans with
+    non-zero maskReason. During the search, at most maxSkip scans with no
+    maskReason set can be skipped. Only ranges with more than minBadScans are
+    kept.
+
+    Args:
+        data: The TimeSeriesData object to pull data from
+        vfat: The VFAT to return ranges for
+        channel: The channel to return ranges for
+        maxSkip: The maximum number of "good" scans between two "bad" scans
+        minBadScans: The minimum number of bad scans
+
+    Returns:
+        A list of MaskedRange objects
+    """
+    ranges = _findRangesMeta(data,
+                             vfat,
+                             channel,
+                             data.maskReason[vfat,channel] != 0,
+                             maxSkip)
+
+    return list(filter(lambda r: r.badMaskReasonScanCount() >= minBadScans,
+                       ranges))
+
+def findRangesMask(data, vfat, channel, maxSkip = 5, minBadScans = 4):
+    """Finds ranges of scans based on the mask attribute.
+
+    Searches the data for the given vfat and channel for ranges of scans with
+    non-zero maskReason. During the search, at most maxSkip scans with mask not
+    set can be skipped. Only ranges with more than minBadScans are kept.
+
+    Args:
+        data: The TimeSeriesData object to pull data from
+        vfat: The VFAT to return ranges for
+        channel: The channel to return ranges for
+        maxSkip: The maximum number of "good" scans between two "bad" scans
+        minBadScans: The minimum number of bad scans
+
+    Returns:
+        A list of MaskedRange objects
+    """
+    ranges = _findRangesMeta(data,
+                             vfat,
+                             channel,
+                             data.mask[vfat,channel] != 0,
+                             maxSkip)
+
+    return list(filter(lambda r: r.maskedScanCount() >= minBadScans,
+                       ranges))
+
+def findRangesNoise(data,
+                    vfat,
+                    channel,
+                    threshold = 0.02,
+                    maxSkip = 5,
+                    minBadScans = 4):
+    """Finds ranges of scans based on noise.
+
+    Searches the data for the given vfat and channel for ranges of scans with
+    noise < threshold. During the search, at most maxSkip scans with mask not
+    set can be skipped. Only ranges with more than minBadScans are kept.
+
+    Args:
+        data: The TimeSeriesData object to pull data from
+        vfat: The VFAT to return ranges for
+        channel: The channel to return ranges for
+        threshold: The minimum value of noise for a scan to be "good"
```

Comment:
What is "good" defined as?  Additionally `threshold` is exactly the name of the `TBranch` which tracks scurve_mean value but here it is related to `scurve_sigma`.  Change to be something less confusing.  e.g. `scurveSigmaCut` or `noiseCut` or similar.